File: tests/test_integrations/test_langchain/apps/langchain_metric_collection_app.py
"""
Metric Collection LangChain App: Tests metric_collection on LLM and tool spans
Complexity: LOW - Tests metric_collection tracing

Uses ChatOpenAI with metric_collection in metadata and the patched @tool decorator
with metric_collection for component-level evaluations.
"""

from langchain.agents import create_agent
from langchain_openai import ChatOpenAI
from langchain_core.runnables import RunnableConfig

from deepeval.integrations.langchain import tool
from deepeval.prompt import Prompt

# Create a Prompt object for prompt tracking
test_prompt = Prompt(alias="metric-collection-test-prompt")
test_prompt.version = "01.00.00"


@tool(metric_collection="tool_accuracy")
def calculate(expression: str) -> str:
    """Evaluates a simple math expression and returns the result."""
    # Simple calculator that handles basic operations
    try:
        # Only allow safe characters
        allowed = set("0123456789+-*/.(). ")
        if not all(c in allowed for c in expression):
            return "Error: Invalid characters in expression"
        result = eval(expression)
        return str(result)
    except Exception as e:
        return f"Error: {str(e)}"


# LLM with metric_collection and prompt in metadata
llm = ChatOpenAI(
    model="gpt-5-mini",
    temperature=0,
    seed=42,
    metadata={
        "metric_collection": "llm_quality",
        "prompt": test_prompt,
    },
)

# Create agent with the tool
agent_executor = create_agent(
    llm,
    [calculate],
    system_prompt="You are a calculator assistant. Use the calculate tool to evaluate math expressions.",
)


def invoke_metric_collection_app(inputs: dict, config: RunnableConfig = None):
    """Invoke the metric collection app."""
    return agent_executor.invoke(inputs, config=config)


async def ainvoke_metric_collection_app(
    inputs: dict, config: RunnableConfig = None
):
    """Async invoke the metric collection app."""
    return await agent_executor.ainvoke(inputs, config=config)

File: typescript/test/test-integrations/test-langchain/apps/langchain-metric-collection-app.ts
import { tool } from "@langchain/core/tools";
import { z } from "zod";
import { ChatOpenAI } from "@langchain/openai";
import { RunnableLambda, RunnableConfig } from "@langchain/core/runnables";
import {
  BaseMessage,
  AIMessage,
  ToolMessage,
  SystemMessage,
} from "@langchain/core/messages";
import { Prompt } from "../../../../src/prompt";

const testPrompt = new Prompt({ alias: "langchain-metric-collection-test" });
testPrompt.version = "01.00.00";
testPrompt.label = "production";
testPrompt.hash = "bab04ec";

const calculate = tool(
  (input) => {
    try {
      const allowed = new Set("0123456789+-*/.(). ".split(""));
      if (![...input.expression].every((c) => allowed.has(c)))
        return "Error: Invalid characters";
      return String(eval(input.expression));
    } catch (e: any) {
      return `Error: ${e.message}`;
    }
  },
  {
    name: "calculate",
    description: "Evaluates a math expression.",
    schema: z.object({ expression: z.string() }),
  },
);

const llm = new ChatOpenAI({
  modelName: "gpt-4o-mini",
  temperature: 0,
  metadata: {
    prompt: testPrompt,
    metricCollection: "llm-span-evals",
  },
});

const llmWithTools = llm.bindTools([calculate]);

const runMetricCollectionApp = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => {
  const systemMsg = new SystemMessage("You are a calculator assistant.");
  const allMessages = [systemMsg, ...(inputs.messages || [])];

  for (let i = 0; i < 2; i++) {
    const response = (await (llmWithTools.invoke as any)(
      allMessages,
      config,
    )) as AIMessage;
    allMessages.push(response);

    if (!response.tool_calls || response.tool_calls.length === 0) break;

    for (const toolCall of response.tool_calls) {
      if (toolCall.name === "calculate") {
        const result = await calculate.invoke(toolCall, config);
        allMessages.push(
          new ToolMessage({
            content:
              typeof result === "string" ? result : JSON.stringify(result),
            tool_call_id: toolCall.id!,
          }),
        );
      }
    }
  }
  return { messages: allMessages.slice(1) };
};

const metricCollectionChain = new RunnableLambda({
  func: runMetricCollectionApp,
}).withConfig({ runName: "metric_collection_chain" });

export const invokeMetricCollectionApp = async (
  inputs: { messages: BaseMessage[] },
  config?: RunnableConfig,
) => {
  return await metricCollectionChain.invoke(inputs, config);
};
